File: prueba.py
from abc import ABC, abstractmethod

class Component(ABC):
    @abstractmethod
    def render(self):
        pass
class Button(Component):
    def render(self):
        return "<button> </button>"

class InputText(Component):
    def render(self):
        return "<input />"

class Select(Component):
    def render(self):
        return "<select> </select>"

# ==============================
# PATRÓN DECORATOR
# ==============================
class TypeDecorator(Component):
    def __init__(self, component):
        self.component = component

    def render(self):
        return self.component.render()
    
class InputTypeDecorator(TypeDecorator):
    def __init__(self, component):
        super().__init__(component)

    def render(self):
        html = self.component.render()
        html = html.replace("<input", '<input type="radio"')
        return html

# ==============================
# PATRÓN COMPOSITE
# ==============================
class Formulario(Component):  # Composite
    def __init__(self):
        self.components = []

    def add_component(self, comp):
        self.components.append(comp)

    def render(self):
        html = ""
        for comp in self.components:
            html += comp.render()
        return f"<div>{html}</div>"

# ==============================
# PATRÓN ADAPTER
# ==============================
class SalidaAdapter:
    def __init__(self, obj):
        # Convertir a Component si es necesario
        if isinstance(obj, Component):
            self.obj = obj
        elif isinstance(obj, EtiquetaTexto):
            self.obj = EtiquetaTextoAdapter(obj)
        else:
            self.obj = None

    def getEntradaFormada(self):
        if self.obj:
            return self.obj.render()
        else:
            return "<!-- componente no compatible -->"

# Componente externo no compatible originalmente
class EtiquetaTexto:
    def generar_etiqueta(self):
        return f"<label> </label>"

# Adaptador para EtiquetaTexto
class EtiquetaTextoAdapter(Component):
    def __init__(self, etiqueta_texto):
        self.etiqueta_texto = etiqueta_texto

    def render(self):
        return self.etiqueta_texto.generar_etiqueta()

# ==============================
# PATRÓN SINGLETON
# ==============================
class Render:
    _instance = None

    def __init__(self):
        if Render._instance is not None:
            raise Exception("Esta clase es Singleton. Usa getInstance()")
        Render._instance = self

    @staticmethod
    def getInstance():
        if Render._instance is None:
            Render()
        return Render._instance

    def render_html(self, contenido):
        return f"<html><body><form>{contenido}</form></body></html>"

# ==============================
# APLICACIÓN PRINCIPAL
# ==============================
class Application:
    @staticmethod
    def main():

        # Componente externo + adaptador
        etiqueta_externa = EtiquetaTexto()
        adaptador_etiqueta = EtiquetaTextoAdapter(etiqueta_externa)
        # Crear formulario principal
        formulario_principal = Formulario()

        # Crear formulario anidado
        formulario_anidado = Formulario()
        formulario_anidado.add_component(Button())
        formulario_anidado.add_component(InputText())
        formulario_anidado.add_component(adaptador_etiqueta)

        # Agregar formulario anidado dentro del principal
        formulario_principal.add_component(formulario_anidado)

        # Agregar componentes simples al formulario principal
        formulario_principal.add_component(Select())
        formulario_principal.add_component(InputText())

        # Usar el adaptador y render singleton como antes
        adaptador = SalidaAdapter(formulario_principal)
        salida_html = adaptador.getEntradaFormada()

        render = Render.getInstance()
        html_completo = render.render_html(salida_html)
        return html_completo

# Ejecutar aplicación
print(Application.main())

File: VariosPatrones/prueba.swift
protocol Component {
    func render() -> String
}

class Button: Component {
    func render() -> String {
        return "<button> </button>"
    }
}

class InputText: Component {
    func render() -> String {
        return "<input />"
    }
}

class Select: Component {
    func render() -> String {
        return "<select> </select>"
    }
}

// Decorator
class InputTypeDecorator: Component {
    
    private var component: Component

    init(component: Component) {
        self.component = component
    }

   func render() -> String {
        let original = component.render()
        return original.replacing("<input ", with: "<input type='radio'")
    }
}

// Composite
class FormBody: Component {
    private var components: [Component] = []

    func addComponent(_ comp: Component) {
        components.append(comp)
    }

    func render() -> String {
        var formBody = ""
        for comp in components {
            formBody += comp.render()
        }
        return "<div>" + formBody + "</div>"
    }
}

class Formulario: Component {
    private var body: FormBody

    init(body: FormBody) {
        self.body = body
    }

    func render() -> String {
        return "<form>" + body.render() + "</form>"
    }
}

// Adapter
class EtiquetaTexto {
    func generarEtiqueta() -> String {
        return "<label> </label>"
    }
}

class EtiquetaTextoAdapter: Component {
    
    private var etiquetaTexto: EtiquetaTexto

    init(etiquetaTexto: EtiquetaTexto) {
        self.etiquetaTexto = etiquetaTexto
    }

    func render() -> String {
        return etiquetaTexto.generarEtiqueta()
    }
}


// Singleton
class Render {
    static var instance: Render? = nil

    private init() {}

    static func getInstance() -> Render? {
        if instance == nil {
            instance = Render()
        }
        return instance
    }

    func renderHTML(listaDeFormularios: [Component]) -> String {
        var contenidoForms = ""
        for form in listaDeFormularios {
            contenidoForms += form.render()
        }
        return "<html><body>" + contenidoForms + "</body></html>"
    }
}


let formBody1 = FormBody()
formBody1.addComponent(Button())
formBody1.addComponent(InputTypeDecorator(component: InputText()))

let formBody2 = FormBody()
formBody2.addComponent(Select())
formBody2.addComponent(EtiquetaTextoAdapter(etiquetaTexto: EtiquetaTexto()))
formBody2.addComponent(formBody1)

let form1 = Formulario(body: formBody2)

if let pagina = Render.getInstance() {
    let htmlFinal = pagina.renderHTML(listaDeFormularios: [form1])
    print(htmlFinal)
}
